### DISCUSSION TITLE
Aggressive Front End Caching

### DISCUSSION THREAD

**Discussion by Azza-Aaron** [serwist/serwist#137] [Q&A] [answered]
Hi,

I've got an app using your next-pwa modification, I would like to swap to Serwist but wanted to check if there was still an option to apply Aggressive front end caching. Currently as I swap to Serwist it doesn't seem to cache assets that haven't been visited.

My current config is as per your documentation, also are we still able to change the approach - online first/stale while revalidate?

Thanks for your help! (I read through your documentation, forgive me if I missed anything!)

**Comment by DuCanhGH** [comment, marked as answer]
@Azza-Aaron 

> I've got an app using your next-pwa modification, I would like to swap to Serwist but wanted to check if there was still an option to apply Aggressive front end caching. Currently as I swap to Serwist it doesn't seem to cache assets that haven't been visited.

Well, that's runtime caching for you :) If you want to manually cache some assets before they are visited, you can tell Serwist to cache them for you by sending it a message:

```ts
window.serwist.messageSW({
  type: "CACHE_URLS",
  payload: { urlsToCache },
});
```

Usually you'd precache assets your users will often use though.

> My current config is as per your documentation, also are we still able to change the approach - online first/stale while revalidate?

Not sure what this means. You sure can use `NetworkFirst` and `StaleWhileRevalidate` if you want to. Here are a few resources you may want to look into:

- [Serwist's documentation on its routing system](https://serwist.pages.dev/docs/serwist/runtime-caching/routing)

- [Serwist's documentation on caching strategies](https://serwist.pages.dev/docs/serwist/runtime-caching/caching-strategies)

`aggressiveFrontEndNavCaching` does not exist in Serwist. In hindsight, I'd say that `cacheOnFrontEndNav` and `aggressiveFrontEndNavCaching` were some sort of design mistakes. As soon as the browser requests for an asset, the request is intercepted by the service worker, and if the response hasn't been cached, the selected strategy _may_ cache it for you. This depends on the strategy's implementation though. In theory, we don't actually need a separate worker that manually parses HTML responses and caches their assets, rendering the two aforementioned options obsolete.

**Reply by Azza-Aaron** [reply]
Thanks very much for your detailed response!
